### DISCUSSION TITLE
Airflow DAG with PythonOperators module dependency conflict

### DISCUSSION THREAD

**Discussion by besar22** [apache/airflow#24522] [Q&A]
Hello, 

I am using airflow dag to automate a complex pipeline. I have custom code structure build into my project that defines airflow task metadata. The metadata consist of  the python callable, custom arguments and the dependencies on other python callables. This means, my project is imported into the dag file and the tasks and dependencies are automatically generated.  Some code is run inside my project to get the logic for the input arguments of the tasks required for airflow's PythonOperator.

The problem is that my project and airflow have to run on the same python environment but they have dependency problems. (E.g. sqlalchemy, my project requires an older version). 

What I think I have found so far:
- PythonVirtualenvOperator does not solve my issue because 1- my project has many dependencies 2- the tasks cannot be constructed as the dag file cannot run the project code.
- my project's dependencies cannot be included in the module, i.e. then pip would have to install 1 module only. I also need the code of the module to be imported
- Airflow Dag File Processor cannot run as a separate process inside a separate python environment 

Are my finding correct? What suggestions do you have to solve this problem? Can it be solved at all?
Any suggestions are welcomed.  
Thanks

**Comment by potiuk** [comment]
Yes. If you want to run your code as Python callable using Python Operator (and all Python operators) then you have to make sure your deps are not conflicting. If you do not want to run virtualenv operator, there are two more choiices - you can use DockerOperator or KubernetesPod Operator. Then you can prepare the image with yuor dependencies and run your task as a command using Docker container or as Kubernetes Pod.

**Comment by potiuk** [comment]
You can even (in Airflow 2.3) use @task.docker decorator - which means that your Python code to exacute can use local imports and import libraries that are present in your image, And the @task.docker decorator will serialize your function and send it to the docker container to execute - this is the best you can do I think.

**Comment by besar22** [comment]
@potiuk Many thanks for your reply. 

At the moment, I have custom code that has to run during task creations (i.e. the dag file) that generates custom ranges for task parallelization. If I use the docker container approach (this solves the conflicting dependencies I believe) then I loose the flexibility and I have to hardcode the structure. I use airflow environment variables to ask my code to generate parallel tasks. E.g. I need 4 parallel workers and my code spits out the a list of keyword arguments, which are then used to call the same python callable.  

Its the same problem I have with the PythonVirtualenvOperator as I am willing to suffer the time penalty as usually my task are from several minutes to several hours. Docker / Kubernetes does not work in this case. 

Is there any way to "precompile" all python dependencies such that my module uses its own versions instead of what pip installs? This makes logic sense in theory. Is there a reason why no such mechanism exists?

**Reply by potiuk** [reply]
You can use https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-42+Dynamic+Task+Mapping which is released in Airflow 2.3 to solve your paralellisation problem. Don't use Airflow Variables for that.

> Is there any way to "precompile" all python dependencies such that my module uses its own versions instead of what pip installs? This makes logic sense in theory. Is there a reason why no such mechanism exists?

This is limitation of Python not Airflow.
